modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/statechannel/StateChannelValidator.scala

```diff
@@ -34,19 +34,43 @@ trait StateChannelValidator[F[_]] {
   def validate(
     stateChannelOutput: StateChannelOutput,
     snapshotOrdinal: SnapshotOrdinal,
-    staked: Balance
+    snapshotFeesInfo: SnapshotFeesInfo
   )(implicit hasher: Hasher[F]): F[StateChannelValidationErrorOr[StateChannelOutput]]
+
   def validateHistorical(
     stateChannelOutput: StateChannelOutput,
     snapshotOrdinal: SnapshotOrdinal,
-    staked: Balance
+    snapshotFeesInfo: SnapshotFeesInfo
   )(
     implicit hasher: Hasher[F]
   ): F[StateChannelValidationErrorOr[StateChannelOutput]]
-
 }
 
 object StateChannelValidator {
+  def getFeeAddresses(info: GlobalSnapshotInfo): Map[Address, Set[Address]] =
+    info.lastCurrencySnapshots.collect {
+      case (address, Right((_, info))) => Map(address -> info.lastMessages.toList.flatMap(_.values.map(_.address).toList).toSet)
+    }.flatten.toMap
```

Comment:
I think that creating an immediate `Map` and flattening is redundant. It should work as well:

```scala
info.lastCurrencySnapshots.collect {
  case (address, Right((_, info))) => address -> info.lastMessages.toList.flatMap(_.values.map(_.address).toList).toSet
}
```

Basically if you return a tuple from `collect` called on `Map` then it stays as `Map`: